system/service/auth_credentials.go

```diff
@@ -0,0 +1,1157 @@
+		}
+	)
+
+	err = func() error {
+		if !svc.settings.Auth.Internal.Enabled || !svc.settings.Auth.Internal.PasswordCreate.Enabled {
+			return AuthErrPasswordCreateDisabledByConfig(aam)
+		}
+
+		if u, err = store.LookupUserByEmail(ctx, svc.store, email); err != nil {
+			return err
+		}
+
+		ctx = internalAuth.SetIdentityToContext(ctx, u)
+
+		if url, err = svc.sendPasswordCreateToken(ctx, u); err != nil {
+			return err
+		}
+
+		return nil
+	}()
+
+	return url, svc.recordAction(ctx, aam, AuthActionGeneratePasswordCreateToken, err)
+}
+
+func (svc *auth) sendPasswordCreateToken(ctx context.Context, u *types.User) (url string, err error) {
+	token, err := svc.createUserToken(ctx, u, credentialsTypeCreatePasswordToken)
+	if err != nil {
+		return
+	}
+
+	return svc.notifications.PasswordCreate(token)
+}
+
+// ValidatePasswordResetToken validates password reset token
+func (svc *auth) ValidatePasswordResetToken(ctx context.Context, token string) (user *types.User, err error) {
+	return svc.loadFromTokenAndConfirmEmail(ctx, token, credentialsTypeResetPasswordToken)
+}
+
+// ValidatePasswordCreateToken validates password create token
+func (svc *auth) ValidatePasswordCreateToken(ctx context.Context, token string) (user *types.User, err error) {
+	return svc.loadFromTokenAndConfirmEmail(ctx, token, credentialsTypeCreatePasswordToken)
+}
+
+// ExchangePasswordResetToken exchanges reset password token for a new one and returns it with user info
+func (svc *auth) ExchangePasswordResetToken(ctx context.Context, token string) (u *types.User, t string, err error) {
+	var (
+		aam = &authActionProps{
+			user:        u,
+			credentials: &types.Credentials{Kind: credentialsTypeResetPasswordToken},
+		}
+	)
+
+	err = func() error {
+		if !svc.settings.Auth.Internal.Enabled || !svc.settings.Auth.Internal.PasswordReset.Enabled {
+			return AuthErrPasswordResetDisabledByConfig(aam)
+		}
+
+		u, err = svc.loadUserFromToken(ctx, token, credentialsTypeResetPasswordToken)
+		if err != nil {
+			return AuthErrInvalidToken(aam).Wrap(err)
+		}
+
+		aam.setUser(u)
+		ctx = internalAuth.SetIdentityToContext(ctx, u)
+
+		t, err = svc.createUserToken(ctx, u, credentialsTypeResetPasswordTokenExchanged)
+		if err != nil {
+			u = nil
+			t = ""
+			return AuthErrInvalidToken(aam).Wrap(err)
+		}
+
+		return nil
+	}()
+
+	return u, t, svc.recordAction(ctx, aam, AuthActionExchangePasswordResetToken, err)
+}
+
+// ChangePassword validates old password and changes it with new
+func (svc *auth) ChangePassword(ctx context.Context, userID uint64, oldPassword, newPassword string) (err error) {
+	var (
+		u  *types.User
+		cc types.CredentialsSet
+
+		aam = &authActionProps{
+			user:        u,
+			credentials: &types.Credentials{Kind: credentialsTypePassword},
+		}
+	)
+
+	err = func() error {
+		if !svc.settings.Auth.Internal.Enabled {
+			return AuthErrInternalLoginDisabledByConfig(aam)
+		}
+
+		if len(oldPassword) == 0 {
+			return AuthErrPasswordNotSecure(aam)
+		}
```

Comment:
This doesn't need to happen; if the password was previously ok and the code below matches to it, this bit doesn't matter